### PR TITLE
Convert property UseNumeric to UseAlphaNumeric

### PR DESCRIPTION
**Pull request by mpmadhavig**
### Proposed changes in this pull request
Adding reverted changes again after rebasing the branch with the upstream master branch.

Related PRs:
- https://github.com/wso2-extensions/identity-governance/pull/777
- https://github.com/wso2-extensions/identity-governance/pull/788

Resolves https://github.com/wso2/product-is/issues/17073

### CODE REVIEW

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 1]
```diff
@@ -233,10 +237,68 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
+     * @return True if the connector is email OTP connector and have both old and new email OTP type related configs.
+     */
+    private boolean isEmailOTPConnectorWithNewConfig(String connectorName, ConnectorConfig connectorConfig) {
+
+        /*
+        If the new config is also added, the length of the properties will be 5 with both old
+        and new OTP type properties. Here the purpose of adding 'at least check' is to allow
+        developers to safely add new configs.
+         */
+        return EMAIL_OTP_AUTHENTICATOR.equals(connectorName) && connectorConfig.getProperties().length >= 5;
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:302]: Need to evaluate this logic. Validating the length of the properties and deciding if it is due to new configs is error-prone.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 2]
```diff
@@ -233,10 +237,68 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
+    /**
+     * This method is used to convert the property value useNumericCharacters to useAlphanumericCharacters.
+     *
+     * @param connectorName   Name of the connector.
+     * @param connectorConfig Connector configuration.
+     */
+    private void convertPropertyUseNumericToUseAlphaNumeric(String connectorName, ConnectorConfig connectorConfig) {
+
+        // Verify the order of the connector properties hasn't changed.
+        if (EMAIL_OTP_USE_ALPHANUMERIC_CHARS.equals(connectorConfig.getProperties()[3].getName()) &&
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:274]: We better check contains rather, evaluate with the position.

Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:274]: Resolved.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java (hunk 1 above)
Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:302]: Resolved

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 3]
```diff
@@ -233,10 +237,68 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
         return null;
     }
 
+    /**
+     * This method is used to make sure property value useNumericCharacters and useAlphanumericCharacters both uses
+     * same user input.
+     *
+     * @param configurationDetails Configuration details of the email OTP connector.
+     */
+    private void convertPropertyUseNumericToUseAlphaNumeric(Map<String, String> configurationDetails) {
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:256]: Change method name

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 4]
```diff
@@ -233,10 +237,68 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
+        }
+    }
+
+    /**
+     * This method is used to convert the property value useNumericCharacters to useAlphanumericCharacters.
+     *
+     * @param connectorName   Name of the connector.
+     * @param connectorConfig Connector configuration.
+     */
+    private void convertPropertyUseNumericToUseAlphaNumeric(String connectorName, ConnectorConfig connectorConfig) {
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:271]: Change method name

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 5]
```diff
@@ -233,10 +237,68 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
+    /**
+     * This method is used to make sure property value useNumericCharacters and useAlphanumericCharacters both uses
+     * same user input.
+     *
+     * @param configurationDetails Configuration details of the email OTP connector.
+     */
+    private void convertPropertyUseNumericToUseAlphaNumeric(Map<String, String> configurationDetails) {
+
+        if (configurationDetails.containsKey(EMAIL_OTP_USE_ALPHANUMERIC_CHARS) &&
+                configurationDetails.containsKey(EMAIL_OTP_USE_NUMERIC_CHARS)) {
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:259]: check only existence of alphanumeric.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java [hunk 6]
```diff
@@ -233,10 +238,83 @@ public ConnectorConfig getConnectorWithConfigs(String tenantDomain,
...
+
+        if (configurationDetails.containsKey(EMAIL_OTP_USE_ALPHANUMERIC_CHARS)) {
+            // Gets the value of alphanumeric property value and negates it.
+            boolean useNumericChars = !Boolean.parseBoolean(configurationDetails.get(EMAIL_OTP_USE_ALPHANUMERIC_CHARS));
+            configurationDetails.put(EMAIL_OTP_USE_NUMERIC_CHARS, String.valueOf(useNumericChars));
+        }
+    }
+
+    /**
+     * This method is used to convert the property value useNumericCharacters to useAlphanumericCharacters.
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:266]: for email OTP connector.

Comment (hwupathum) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImpl.java:266]:
```suggestion
     * This method is used to convert the property value useNumericCharacters to useAlphanumericCharacters for email OTP connector.
```

**Review by hwupathum — APPROVED** [review]

**Pull request merged**